src/pybind/mgr/cephadm/services/monitoring.py

```diff
@@ -423,6 +423,20 @@ def config_dashboard(self, daemon_descrs: List[DaemonDescription]) -> None:
                 service_url
             )
 
+    def pre_remove(self, daemon: DaemonDescription) -> None:
+        """
+        Called before Alertmanager is removed
+        """
+        if daemon.hostname is not None:
+            try:
+                self.mgr.check_mon_command({"prefix": "dashboard reset-alertmanager-api-host"})
+                self.mgr.check_mon_command(
+                    {"prefix": "dashboard reset-alertmanager-api-ssl-verify"}
+                )
```

Comment:
I'm wondering if we want to do this only if we're removing the final alertmanager daemon (we default to deploying one, but nothing is stopping a user from deploying more) or on just any daemon removal. @nizamial09 do you have any thoughts on this? The actual impact of this issue shows up more on the dashboard side rather than within cephadm.

Reply:
I have tested this scenario with Alertmanager deployed on all three nodes of the cluster. When a daemon is removed, the pre_remove function is executed for each removal. Additionally, the orchestrator immediately redeploys the Alertmanager instance, and this change is reflected in the UI as expected.

```
2024-12-05T05:06:57.318422+0000 mgr.ceph-node-0.vxbxjs (mgr.14325) 36547 : cephadm [INF] Remove daemons alertmanager.ceph-node-1
2024-12-05T05:06:57.371909+0000 mgr.ceph-node-0.vxbxjs (mgr.14325) 36550 : cephadm [INF] Reset the Alertmanager values to default
2024-12-05T05:06:57.373371+0000 mgr.ceph-node-0.vxbxjs (mgr.14325) 36551 : cephadm [INF] Removing daemon alertmanager.ceph-node-1 from ceph-node-1 -- ports [9093, 9094]
2024-12-05T05:07:00.804282+0000 mgr.ceph-node-0.vxbxjs (mgr.14325) 36554 : cephadm [INF] Deploying daemon alertmanager.ceph-node-1 on ceph-node-1

```

To improve it, I can add a check to determine if the daemon being removed is the last one. Does that approach sound good?

Reply:
It might be bit more complicated than that. I see in the alertmanager class' `config_dashboard` function we only set the service url for what we mark as the "active daemon", but the `get_active_daemon`function is just arbitrarily picking the first one in the list. In theory, we'd want to only clear the alertmanager api settings if the alertmanager being removed is the one for we selected as the "active" one when setting the api settings. I think if we find that it consistently picks the same alertmanager for the active one, perhaps we can check if that's the active one and only do this cleanup in that (as said it's arbitrary, but I'm not sure if it's random) or we need to get the current api setting and check if the ip and port for the alertmanager being removed match the setting. In either of those cases, when we do the removal we'd need to clear the current setting and then make sure the api settings get set for one of the remaining alertmanager daemons. On that note I'll add that while the alertmanager daemon gets replaced in the case you just run a `ceph orch daemon rm...`, that won't happen if the daemon gets removed as the result of a change in placement. E.g. if you go from
```
placement:
  hosts:
   - ceph-node-0
   - ceph-node-1
   - ceph-node-2
```
to
```
placement:
  hosts:
   - ceph-node-1
   - ceph-node-2
```
the alertmanager on `ceph-node-0` will be removed and will NOT be replaced. That case is actually closer to what I'd expect to happen in a real environment. There's not often reason to run a `ceph orch daemon rm..` and users scaling down an alertmanager service seems more likely. So you'd have to factor in the fact that no new alertmanager daemons may be deployed when doing the implementation. All of this is of course assuming we have multiple alertmanager daemons to begin with. If we see we're removing the only one we can just go forward with removing the setting as you're doing currently

Reply:
I've known some set of users deploy multiple number of alertmanager (and other monitoring stacks) for the purpose of HA. Even if the number is very limited it is a possibility so I guess if we need to reset the config it makes sense to reset it only when the active instance gets removed. If we remove it for every daemon, the URL disappears and will leave the UI from not recieving any alerts from the alertmanager. I guess the same goes for prometheus as well.

Reply:
The scenarios below work correctly for Alertmanager, but when repeated for Prometheus, removing the active daemon causes the Dashboard to show a 404 error. However, the CLI indicates that the API settings have been updated to point to the available daemon.

The CLI logs confirm that the dashboard set-prometheus-api-host has been updated to point to the remaining daemon (http://ceph-node-1.local:9095), but the Dashboard fails to connect.

```
404 - Not Found
Could not reach Prometheus's API on http://ceph-node-1.local:9095/api/v1 error HTTPConnectionPool(host='ceph-node-1.local', port=9095): Max retries exceeded with url: /api/v1/rules (Caused by NewConnectionError(': Failed to establish a new connection: [Errno -2] Name or service not known'))
```


**Alertmanager Scenario 1: Removing the daemon which is not active.** 
```
[ceph: root@ceph-node-0 ~]# ceph orch apply alertmanager --placement '*'
Scheduled alertmanager update...

[ceph: root@ceph-node-0 ~]# ceph orch ls
NAME                       PORTS        RUNNING  REFRESHED  AGE  PLACEMENT  
alertmanager               ?:9093,9094      3/3  31s ago    91s  *          
crash                                       3/3  31s ago    11m  *          
mgr                                         2/2  31s ago    11m  count:2    
mon                                         3/5  31s ago    11m  count:5    
node-exporter              ?:9100           3/3  31s ago    3m   *          
osd.all-available-devices                     3  31s ago    10m  * 

[ceph: root@ceph-node-0 ~]# ceph dashboard get-alertmanager-api-host
http://ceph-node-0.local:9093

[ceph: root@ceph-node-0 ~]# ceph orch apply alertmanager --placement="ceph-node-0,ceph-node-1"
Scheduled alertmanager update...

[ceph: root@ceph-node-0 ~]# ceph orch ls 
```
NAME                       PORTS        RUNNING  REFRESHED  AGE  PLACEMENT                
alertmanager               ?:9093,9094      2/2  5s ago     12s  ceph-node-0;ceph-node-1 
```
**Scanerio 2 : Removing the active daemon**

[ceph: root@ceph-node-0 ~]# ceph orch apply alertmanager --placement="ceph-node-1"
Scheduled alertmanager update...

[ceph: root@ceph-node-0 ~]# ceph orch ls
NAME                       PORTS        RUNNING  REFRESHED  AGE  PLACEMENT    
alertmanager               ?:9093,9094      1/1  5s ago     14s  ceph-node-1  

[ceph: root@ceph-node-0 ~]# ceph dashboard get-alertmanager-api-host
http://ceph-node-1.local:9093
```

@adk3798 Would it be appropriate to add a retry mechanism in the pre_remove function to ensure the new Prometheus instance is operational before updating the Dashboard settings? 

I also tried `curl` and it's output shows the API to be available but Dashboard error remains the same.

```
[ceph: root@ceph-node-0 ~]# curl http://192.168.100.101:9095/api/v1/rules
{"status":"success","data":{"groups":[{"name":"PrometheusServer",
```